### PR TITLE
[SCEV] Split collecting and applying rewrite info from loop guards (NFC)

### PR DESCRIPTION
**Pull request by fhahn**
Split off collecting rewrite info from loop guards to collectRewriteInfoFromLoopGuards. This allows users of applyLoopGuards to collect rewrite info once in cases where the same loop guards are applied multiple times.

This is used to collect rewrite info once in howFarToZero, which saves a bit of compile-time:
stage1-O3: -0.04%
stage1-ReleaseThinLTO: -0.02%
stage1-ReleaseLTO-g: -0.04%
stage2-O3: -0.02%
https://llvm-compile-time-tracker.com/compare.php?from=117b53ae38428ca66eaa886fb432e6f09db88fe4&to=4ffb7b2e1c99081ccebe6f236c48a0be2f64b6ff&stat=instructions:u

Notably this improves mafft by -0.9% with -O3, -0.11% with LTO and -0.12% with stage2-O3.

### CODE REVIEW

File: llvm/include/llvm/Analysis/ScalarEvolution.h [hunk 1]
```diff
@@ -1299,8 +1299,17 @@ class ScalarEvolution {
   /// sharpen it.
   void setNoWrapFlags(SCEVAddRecExpr *AddRec, SCEV::NoWrapFlags Flags);
 
+  /// Collect rewrite map for loop guards for loop \p L, together with flags
+  /// indidcating if NUW and NSW can be preserved during rewriting.
+  std::tuple<DenseMap<const SCEV *, const SCEV *>, bool, bool>
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/include/llvm/Analysis/ScalarEvolution.h]: Moved collection and rewriting to a LoopGuards class, thanks!

File: llvm/include/llvm/Analysis/ScalarEvolution.h [hunk 2]
```diff
@@ -1299,8 +1299,26 @@ class ScalarEvolution {
...
+    /// indidcating if NUW and NSW can be preserved during rewriting.
+    static LoopGuards collect(const Loop *L, ScalarEvolution &SE);
+
+    /// Try to apply the collected loop guards to \p Expr
+    const SCEV *rewrite(const SCEV *Expr) const;
+  };
+
   /// Try to apply information from loop guards for \p L to \p Expr.
   const SCEV *applyLoopGuards(const SCEV *Expr, const Loop *L);
+  const SCEV *applyLoopGuards(const SCEV *Expr, const LoopGuards &Gards);
```
Comment (nikic) [inline llvm/include/llvm/Analysis/ScalarEvolution.h:1321]:
```suggestion
  const SCEV *applyLoopGuards(const SCEV *Expr, const LoopGuards &Guards);
```

**Review by nikic — APPROVED** [review]
LGTM

**Review by preames — APPROVED** [review]
LGTM

**Pull request merged**